functions/go/enable-gcp-services/gcpservices/project_service_list.go

```diff
@@ -0,0 +1,230 @@
+		return true
+	}
+	return false
+}
+
+// getServiceRNodes returns generated Service RNodes
+func (ps *ProjectServiceList) getServiceRNodes() ([]*yaml.RNode, error) {
+	var services []*yaml.RNode
+	// get namespace
+	ns := ps.GetIdentifier().Namespace
+	name := ps.GetIdentifier().Name
+	serviceList, err := getServicesList(name, ps.Spec.Services, ps.Spec.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+	// generate service resources
+	for _, s := range serviceList {
+		svc, err := createService(s)
+		if err != nil {
+			return nil, err
+		}
+		// add annotations in ProjectServiceList to each service
+		for k, v := range ps.ObjectMeta.Annotations {
+			if ignoreAnnotation(k) {
+				continue
+			}
+			if _, err := svc.Pipe(yaml.SetAnnotation(k, v)); err != nil {
+				return nil, err
+			}
+		}
+		// add management annotation for state tracking
+		if _, err := svc.Pipe(yaml.SetAnnotation(managedByAnnotation, name)); err != nil {
+			return nil, err
+		}
+		// add namespace to service if present
+		if ns != "" {
+			if _, err := svc.Pipe(yaml.SetK8sNamespace(ns)); err != nil {
+				return nil, err
+			}
+		}
+		// add path annotation to generate service in same dir as parent project service list
+		svcMeta, err := svc.GetMeta()
+		if err != nil {
+			return nil, err
+		}
+		serviceFilePath := kioutil.CreatePathAnnotationValue(path.Dir(ps.ObjectMeta.Annotations[kioutil.PathAnnotation]), svcMeta)
+		if _, err := svc.Pipe(yaml.SetAnnotation(kioutil.PathAnnotation, serviceFilePath)); err != nil {
+			return nil, err
+		}
+
+		services = append(services, svc)
+		err = ps.addResult("generated service", svc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return services, nil
+}
+
+// ignoreAnnotation checks if an annotation on ProjectServiceList should be ignored
+func ignoreAnnotation(k string) bool {
+	_, ok := ignoreAnnotations[k]
+	return ok
+}
+
+// addResult records the action performed on a resource
+func (ps *ProjectServiceList) addResult(action string, r *yaml.RNode) error {
+	meta, err := r.GetMeta()
+	if err != nil {
+		return err
+	}
+	fp, _, err := kioutil.GetFileAnnotations(r)
+	if err != nil {
+		return err
+	}
+	ps.results = append(ps.results, Result{meta.GetIdentifier(), fp, action})
+	return nil
+}
+
+// GetResults returns operations performed by ProjectServiceListRunner
+func (r *ProjectServiceListRunner) GetResults() []Result {
+	return r.results
+}
+
+// isManaged checks is a resource is managed by a the ProjectServiceList
+func (ps ProjectServiceList) isManaged(r *yaml.RNode) bool {
+	anno := r.GetAnnotations()
+	k, found := anno[managedByAnnotation]
+	// if no managed annotation found it is not managed
+	if !found {
+		return false
+	}
+	return k == ps.Name
+}
+
+// pruneGeneratedServices removes resources managed by ProjectServiceList
+func (ps *ProjectServiceList) pruneGeneratedServices(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	n := 0
+	for i := range nodes {
+		if !ps.isManaged(nodes[i]) {
+			nodes[n] = nodes[i]
```

Comment:
It seems there is an issue with in-place modification: suppose there are some services managed by ProjectServiceList 1 and some other services are managed by ProjectServiceList 2. If we process ProjectServiceList 1 first, the services managed by ProjectServiceList 2 will be lost.

We should have a test to exercise the behavior

Reply:
Added a test. For a fn invocation all services managed by any ProjectServiceList is pruned. In your example if `ProjectServiceList 2` CR still exists, the objects are regenerated so after fn execution it should be noop.